### DISCUSSION TITLE
could provenance be used to link published forks to their source?

### DISCUSSION THREAD

**Discussion by panva** [npm/feedback#909] [Registry]
Since publishing with provenance checks the package.json repository is the one that is being published from, could this be used to identify forks and label them accordingly when a package is published without provenance pointing to the same repo that another package with provenance has been published from?

This would be a helpful indicator to developers who landed on a forked package page (either through social engineering attacks or an unfortunate search result ordering from <search engine> or npm itself)

e.g.

- repository and homepage is hidden (possibly even removed from the package.json in registry), replaced with link to the original package on the package page.
- funding is hidden (and removed from the package.json in registry?)
- issues and pr count is hidden (or simply not present because the link was removed)

Put in practice, on a low effort fork that has no intention to further maintain or continue publishing upstream fixes:

<details><summary>Before</summary>
<p>

![Screenshot 2023-04-20 at 15 35 34](https://user-images.githubusercontent.com/241506/233383210-d48719b8-7842-4516-8e09-be21fdc3b086.png)

</p>
</details> 

<details><summary>After</summary>
<p>

![image](https://user-images.githubusercontent.com/241506/233381015-04e60211-9a83-46e7-bdd1-ac08148c58d9.png)

</p>
</details> 

This does not prohibit forks to be re-published, but when done so with no effort to updating the README/package.json it avoids showing details that are incorrect/misleading.

Past related threads: #103, #84

**Comment by ljharb** [comment]
This seems like something that should be achievable with or without provenance.

Also, sometimes a fork and a source are both legitimately published - why would either one but not the other using provenance tell you anything about legitimacy?

**Reply by panva** [reply]
> This seems like something that should be achievable with or without provenance.

It seems to me like that is the enabling factor since without it there's no way to prove ownership (at the time that's what #103 came down to).

> Also, sometimes a fork and a source are both legitimately published - why would either one but not the other using provenance tell you anything about legitimacy?

Legitimacy is not my concern per se. Identifying low effort / workaround forks that get published to the registry without the intention to be maintained or actually signinificantly branch out is. These links that get republished without being owned are followed by stray developers who then run into various issues and either abandon using the package (when in fact using the source one would help), or worse, consume maintainer's brain cycles by using an unmaintained fork and then venturing to the issue trackers with cryptic issues that stem from them not even using the right package.

Of course a fork can be there for a good reason and is maintained, in which case (e.g. https://www.npmjs.com/package/npm-run-all and https://www.npmjs.com/package/npm-run-all2) the fork will change the repository and homepage.

**Reply by ljharb** [reply]
npm accounts can link to github accounts, so npm already has the means to verify that the package publisher has Write permissions or higher on the linked repo.

**Reply by panva** [reply]
> npm accounts can link to github accounts, so npm already has the means to verify that the package publisher has Write permissions or higher on the linked repo.

I didn't catch that, cool, then indeed this seems to be achievable using multiple means.

**Reply by panva** [reply]
Altho it would seem that there are plans for provenance to be supported from other platforms than GH/GHA in which case provenance can be used for those. Unless of course, account linking is to be added for those other platforms too.
